sdk/src/storage/preview_storage_client_builder.rs

```diff
@@ -0,0 +1,84 @@
+use momento_protos::control_client::scs_control_client::ScsControlClient;
+use momento_protos::store::store_client::StoreClient;
+use tonic::codegen::InterceptedService;
+
+use crate::grpc::header_interceptor::HeaderInterceptor;
+use crate::storage::{Configuration, PreviewStorageClient};
+use crate::{utils, CredentialProvider, MomentoResult};
+
+pub struct PreviewStorageClientBuilder<State>(pub State);
+
+pub struct NeedsConfiguration;
+
+pub struct NeedsCredentialProvider {
+    configuration: Configuration,
+}
+
+pub struct ReadyToBuild {
+    configuration: Configuration,
+    credential_provider: CredentialProvider,
+}
+
+impl PreviewStorageClientBuilder<NeedsConfiguration> {
+    pub fn configuration(
+        self,
+        configuration: impl Into<Configuration>,
+    ) -> PreviewStorageClientBuilder<NeedsCredentialProvider> {
+        PreviewStorageClientBuilder(NeedsCredentialProvider {
+            configuration: configuration.into(),
+        })
+    }
+}
+
+impl PreviewStorageClientBuilder<NeedsCredentialProvider> {
+    pub fn credential_provider(
+        self,
+        credential_provider: CredentialProvider,
+    ) -> PreviewStorageClientBuilder<ReadyToBuild> {
+        PreviewStorageClientBuilder(ReadyToBuild {
+            configuration: self.0.configuration,
+            credential_provider,
+        })
+    }
+}
+
+impl PreviewStorageClientBuilder<ReadyToBuild> {
+    pub fn build(self) -> MomentoResult<PreviewStorageClient> {
+        let agent_value = &utils::user_agent("sdk");
+
+        let data_channel = utils::connect_channel_lazily_configurable(
+            &self.0.credential_provider.cache_endpoint,
```

Comment:
will probably be storage_endpoint, yeah?

Reply:
Good catch. I'll update that.